homeassistant/components/decora_wifi/config_flow.py

```diff
@@ -0,0 +1,103 @@
+"""Config flow for myLeviton decora_wifi."""
+
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.data_entry_flow import FlowResult
+
+from .common import CommFailed, DecoraWifiPlatform, LoginFailed
+from .const import CONF_TEMPORARY, CONF_TITLE, DOMAIN
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class DecoraWifiConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a Decora Wifi config flow."""
+
+    VERSION = 1
+
+    def __init__(self):
+        """Initialize the config flow class."""
+        self.data = {}
+        self._finish_step = None
+        super().__init__()
+
+    async def async_step_user(self, user_input=None) -> FlowResult:
+        """Handle user-initiated setup config flow."""
+        # Begin interactive credential input
+        if not user_input:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=vol.Schema(
+                    {
+                        vol.Required(CONF_USERNAME): str,
+                        vol.Required(CONF_PASSWORD): str,
+                    }
+                ),
+            )
+        # Update data
+        self.data.update(user_input)
+        # Conduct pre-validation checks
+        await self.async_set_unique_id(self.data[CONF_USERNAME].lower())
+        self._abort_if_unique_id_configured()
+        # Set finish_step and proceed to validation
+        self._finish_step = self.async_step_user_finish
+        return await self.async_step_validate(None)
+
+    async def async_step_validate(self, user_input=None) -> FlowResult:
+        """Call the myLeviton API to validate user-provided credentials."""
+        if user_input:
+            self.data.update(user_input)
+        session = None
+        errors = {}
+        try:
+            session = await DecoraWifiPlatform.async_setup_decora_wifi(
+                self.hass,
+                email=self.data[CONF_USERNAME],
+                password=self.data[CONF_PASSWORD],
+            )
+        except LoginFailed:
+            errors["base"] = "invalid_auth"
+        except CommFailed:
+            errors["base"] = "cannot_connect"
+        if errors:
+            # Show the form with error message and direct the flow back to the retry step configured earlier
+            return self.async_show_form(
+                step_id="validate",
+                data_schema=vol.Schema(
+                    {vol.Required(CONF_PASSWORD, default=self.data[CONF_PASSWORD]): str}
+                ),
+                errors=errors,
+            )
+        # Login validated. Save the session in temp, then move on to the finish step.
+        self.hass.data.setdefault(DOMAIN, {})
+        self.hass.data[DOMAIN][CONF_TEMPORARY] = session
```

Comment:
We don't want to pass data between the config flow and the rest of the integration with `hass.data`. Is there a reason that you can't login with a new session after its been setup?

Reply:
frenck asked in a previous review on this PR that I avoid re-logging in multiple times where possible. I think this is because it reduces traffic to the Leviton servers, and being good citizens in how we access this quasi-public API helps ensure that Leviton doesn't come up with a technical means to kick us off of it.  I can see the argument for accepting one more API login call integration setup in exchange for the simpler code flow though, if that's what you'd prefer.

Reply:
Its definitely good to minimize the load their servers.  We don't want to store the data in hass.data (There is also the problem that two instances being setup in the same time will override the data) here though so I think its better to login once the integration is setup instead for this one specific case.